File: Modifica_km.cc
#include <iostream>
#include <fstream>
#include <math.h>
using namespace std;

//lettura del chilometraggio totale
int lettura(const int &inizio,const int &blocchi,istream &i){
	int valore=0;
	char lett=0;
	i.seekg(inizio);
	for(int x=0;x<blocchi;x++){
		lett=i.get();
		valore+=lett*pow(10,x);
	}
	return valore;
}

//scrittura del nuovo chilometraggio totale
int scrittura(const int &inizio,const int &blocchi,int km,ostream &o){
	int valore=0;
	o.seekp(inizio);
	for(int x=0;x<blocchi;x++){
		valore=km%10;
		o.put(valore);
		km/=10;
	}
	return 0;
}

//programma per modifica km totali in aprilia caponord e simili
int main(int argc,char **argv){
	
	if(argc!=3){
                cerr<<"Usage: "<<argv[0]<<" file_lettura km_nuovi"<<endl;
                return 1;
        }
	
	const int inizio=0x1C; //indirizzo primo byte odometro totale
	const int blocchi=6; //numero di celle
	int km=atoi(argv[2]); //(atoi converte string in int) il secondo argomento che rappresenta i chilometri da inserire viene convertito in int
	char c;

	ifstream fi(argv[1]); //il primo argomento è il nome del file da modificare
	if(!fi)
		cout<<"Errore apertura file input!!"<<endl;
	ofstream fo("output");
	if(!fo)
		cout<<"Errore apertura file output!!"<<endl;

	for(int i=0;i<0x1c;i++){ //copio il file fino al punto da modificare
		fi.get(c);
		fo<<c;
	}
	
	cout<<"Valore attuale: "<<lettura(inizio,blocchi,fi)<<endl; //leggo il valore originale
	if(scrittura(inizio,blocchi,km,fo)!=0)						//scrivo il valore inserito nel nuovo file
		cout<<"Errore durante la scrittura sul file"<<endl;		//e controllo che l'operazione sia andata a buon fine
	
	fi.seekg(inizio+blocchi); //mi sposto a dopo il chilometraggio
	fo.seekp(inizio+blocchi); //in entrambi i file...
	while(fi.get(c)) //...e copio tutto il resto fino alla fine
		fo<<c;

	fi.close();
	fo.close();
	
	ifstream contr("output"); //apro il file creato in lettura
	if(!contr)				//per controllare che tutto sia andato a buon fine
		cout<<"Errore apertura controllo!!"<<endl;
	cout<<"Lettura nuovo valore: "<<lettura(inizio,blocchi,contr)<<endl;
	contr.close();
	
	cout<<"Programma terminato correttamente"<<endl;
	return 0;
}

File: Modifica_km.py
# -*- coding: utf-8 -*-

# lettura del chilometraggio totale
def Leggi(inizio,blocchi,lettura):
    valore = 0
    lettura.seek(inizio)
    for x in range(blocchi):
        lett = lettura.read(1)
        lett = ord(lett) # ord() converte da stringa Unicode a int
        valore += lett*(10**x)
    return valore
    
# scrittura del nuovo chilometraggio totale
def Scrivi(inizio,blocchi,chilometri_nuovo,out):
    valore = 0
    km = chilometri_nuovo
    out.seek(inizio)
    for x in range(blocchi):
        valore = km % 10
        valore = chr(valore)
        out.write(valore)
        km /= 10
   
inizio = 0x1C # primo byte (più significativo) odometro totale
blocchi = 6 # numero totale di byte adibiti al chilometraggio 
in_file = raw_input("inserisci il nome del file da modificare: ")
lettura = open(in_file,'rb')
out = open("out",'wb')
chilometri_nuovo = int(raw_input("inserisci il nuovo numero di chilometri: "))

# copio il file originale
while (True):
    cp = lettura.read(1)
    if cp == '':
        break
    out.write(cp)
 
print( "Valore attuale: %d" % Leggi(inizio,blocchi,lettura))
# modifico nel nuovo file il contenuto dei blocchi dell'odometro totale
Scrivi(inizio,blocchi,chilometri_nuovo,out)

lettura.close()
out.close()

controllo = open("out",'rb')
print("Nuovo valore: %d" % Leggi(inizio,blocchi,controllo))
controllo.close()

